apollo-router/src/services/router/service.rs

```diff
@@ -225,58 +392,48 @@ impl Service<RouterRequest> for RouterService {
 
     fn call(&mut self, req: RouterRequest) -> Self::Future {
         let self_clone = self.clone();
+        let mut this = std::mem::replace(self, self_clone);
 
-        let this = std::mem::replace(self, self_clone);
-
-        let fut = async move { this.call_inner(req).await };
-
-        Box::pin(fut)
+        Box::pin(async move { this.call_inner(req).await })
     }
 }
 
-impl RouterService {
-    async fn process_supergraph_request(
-        self,
-        supergraph_request: SupergraphRequest,
-    ) -> Result<router::Response, BoxError> {
-        // XXX(@goto-bus-stop): This code looks confusing. we are manually calling several
-        // layers with various ifs and matches, but *really*, we are calling each layer in order
-        // and handling short-circuiting.
-        let mut request_res = self
-            .persisted_query_layer
-            .supergraph_request(supergraph_request);
-
-        if let Ok(supergraph_request) = request_res {
-            request_res = self.apq_layer.supergraph_request(supergraph_request).await;
-        }
+impl<S> RouterToSupergraphRequestService<S>
+where
+    S: Service<SupergraphRequest, Response = SupergraphResponse, Error = BoxError> + Send + 'static,
+    S::Future: Send + 'static,
+{
+    async fn call_inner(&mut self, req: RouterRequest) -> Result<RouterResponse, BoxError> {
```

Comment:
This layer is still a bit nasty on the response side: it does some content negotiation, error counting, sets the status code for cancelled requests, and serializes defer and subscription responses. This is fixable but I would rather do it separately, because it is trickier than the stuff I've done here, and it will complicate the PR even more.